File: ruby/lib/runapi/veo_3_1/resources/upscale_video.rb
# frozen_string_literal: true

module RunApi
  module Veo31
    module Resources
      # Veo 3.1 video upscaling resource.
      # Increase resolution of a completed text-to-video or extend-video task to 1080p or 4K.
      class UpscaleVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/veo_3_1/upscale_video"

        RESPONSE_CLASS = Types::UpscaleVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedUpscaleVideoResponse

        def initialize(http)
          @http = http
        end

        # Upscale a video and wait until complete.
        #
        # @param params [Hash] upscale parameters (requires source_task_id and output_resolution)
        # @return [RunApi::Veo31::Types::CompletedUpscaleVideoResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create a video upscale task without waiting.
        #
        # @param params [Hash] upscale parameters
        # @return [RunApi::Veo31::Types::UpscaleVideoResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get video upscale task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Veo31::Types::UpscaleVideoResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          raise Core::ValidationError, "source_task_id is required" unless params[:source_task_id] || params["source_task_id"]
          output_resolution = params[:output_resolution] || params["output_resolution"]
          unless Types::OUTPUT_RESOLUTIONS.include?(output_resolution)
            raise Core::ValidationError, "output_resolution must be one of: #{Types::OUTPUT_RESOLUTIONS.join(", ")}"
          end
        end
      end
    end
  end
end

File: python/src/runapi/veo_3_1/resources/upscale_video.py
"""Veo 3.1 video upscaling resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, ValidationError, RequestOptions

from ..types import (
    OUTPUT_RESOLUTIONS,
    CompletedUpscaleVideoResponse,
    UpscaleVideoResponse,
)


class UpscaleVideo(Resource):
    """Upscale a previously generated video by referencing its task id."""

    ENDPOINT = "/api/v1/veo_3_1/upscale_video"

    RESPONSE_CLASS = UpscaleVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedUpscaleVideoResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Upscale a video and poll until it completes.

        Args:
            **params: video upscale parameters (model, ...).

        Returns:
            The completed (narrowed) video upscale response.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a video upscale task and return immediately with an id.

        Args:
            **params: video upscale parameters (model, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a video upscale task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        if not params.get("source_task_id"):
            raise ValidationError("source_task_id is required")
        output_resolution = params.get("output_resolution")
        if output_resolution not in OUTPUT_RESOLUTIONS:
            raise ValidationError(f"output_resolution must be one of: {', '.join(OUTPUT_RESOLUTIONS)}")
